Applications/Piano/MainWidget.cpp

```diff
@@ -0,0 +1,202 @@
+
+    m_keys_widget = KeysWidget::construct(m_keys_and_knobs_container, audio_engine);
+
+    m_knobs_widget = KnobsWidget::construct(m_keys_and_knobs_container, audio_engine, *this);
+    m_knobs_widget->set_size_policy(SizePolicy::Fixed, SizePolicy::Fill);
+    m_knobs_widget->set_preferred_size(200, 0);
+}
+
+MainWidget::~MainWidget()
+{
+}
+
+// FIXME: There are some unnecessary calls to update() throughout this program,
+// which are an easy target for optimization.
+
+void MainWidget::custom_event(CCustomEvent&)
+{
+    m_wave_widget->update();
+
+    if (m_audio_engine.time() == 0)
+        m_roll_widget->update_roll();
+}
+
+void MainWidget::keydown_event(GKeyEvent& event)
+{
+    // This is to stop held-down keys from creating multiple events.
+    if (m_keys_pressed[event.key()])
+        return;
+    else
+        m_keys_pressed[event.key()] = true;
+
+    note_key_action(event.key(), On);
+    special_key_action(event.key());
+    m_keys_widget->update();
+}
+
+void MainWidget::keyup_event(GKeyEvent& event)
+{
+    m_keys_pressed[event.key()] = false;
+
+    note_key_action(event.key(), Off);
+    m_keys_widget->update();
+}
+
+void MainWidget::note_key_action(int key_code, Switch switch_note)
+{
+    switch (key_code) {
+    case Key_A:
+        m_keys_widget->set_key(0, switch_note);
+        break;
+    case Key_W:
+        m_keys_widget->set_key(1, switch_note);
+        break;
+    case Key_S:
+        m_keys_widget->set_key(2, switch_note);
+        break;
+    case Key_E:
+        m_keys_widget->set_key(3, switch_note);
+        break;
+    case Key_D:
+        m_keys_widget->set_key(4, switch_note);
+        break;
+    case Key_F:
+        m_keys_widget->set_key(5, switch_note);
+        break;
+    case Key_T:
+        m_keys_widget->set_key(6, switch_note);
+        break;
+    case Key_G:
+        m_keys_widget->set_key(7, switch_note);
+        break;
+    case Key_Y:
+        m_keys_widget->set_key(8, switch_note);
+        break;
+    case Key_H:
+        m_keys_widget->set_key(9, switch_note);
+        break;
+    case Key_U:
+        m_keys_widget->set_key(10, switch_note);
+        break;
+    case Key_J:
+        m_keys_widget->set_key(11, switch_note);
+        break;
+    case Key_K:
+        m_keys_widget->set_key(12, switch_note);
+        break;
+    case Key_O:
+        m_keys_widget->set_key(13, switch_note);
+        break;
+    case Key_L:
+        m_keys_widget->set_key(14, switch_note);
+        break;
+    case Key_P:
+        m_keys_widget->set_key(15, switch_note);
+        break;
+    case Key_Semicolon:
+        m_keys_widget->set_key(16, switch_note);
+        break;
+    case Key_Apostrophe:
+        m_keys_widget->set_key(17, switch_note);
+        break;
+    case Key_RightBracket:
+        m_keys_widget->set_key(18, switch_note);
+        break;
+    case Key_Return:
+        m_keys_widget->set_key(19, switch_note);
+        break;
+    }
```

Comment:
This whole `switch` statement sounds like it could be a separate function (say, `key_code_to_key` or something??), and then this should just call `m_keys_widget->set_key(key, switch_note);`

Reply:
So put the switch in a new method called `key_code_to_key` in `KeysWidget` and have `note_key_action` call it?

Reply:
I'm not sure about the name, but yeah. Have `note_key_action()` call it and pass its result to `m_keys-widget->set_key()`. Would be cleaner IMO :slightly_smiling_face:

Reply:
Oh so `note_key_action()` gets an `int` from `key_code_to_key()` and then uses that call `m_keys-widget->set_key()`? I like that